python/src/spark_rapids_ml/umap.py

```diff
@@ -0,0 +1,280 @@
+            "random_state": None,
+            "callback": None,
+            "verbose": False,
+            "output_type": None,
+            "handle": None,
+        }
+
+
+class _UMAPCumlParams(_CumlParams, HasFeaturesCol, HasFeaturesCols, HasLabelCol):
+    def __init__(self) -> None:
+        super().__init__()
+        self._setDefault()
+
+    def getFeaturesCol(self) -> Union[str, List[str]]:  # type: ignore
+        """
+        Gets the value of :py:attr:`featuresCol` or :py:attr:`featuresCols`
+        """
+        if self.isDefined(self.featuresCols):
+            return self.getFeaturesCols()
+        elif self.isDefined(self.featuresCol):
+            return self.getOrDefault("featuresCol")
+        else:
+            raise RuntimeError("featuresCol is not set")
+
+    def setFeaturesCol(self: P, value: Union[str, List[str]]) -> P:
+        """
+        Sets the value of :py:attr:`featuresCol` or :py:attr:`featuresCols`. Used when input vectors are stored in a single column.
+        """
+        if isinstance(value, str):
+            self.set_params(featuresCol=value)
+        else:
+            self.set_params(featuresCols=value)
+        return self
+
+    def setFeaturesCols(self: P, value: List[str]) -> P:
+        """
+        Sets the value of :py:attr:`featuresCols`. Used when input vectors are stored as multiple feature columns.
+        """
+        return self.set_params(featuresCols=value)
+
+    def setLabelCol(self: P, value: str) -> P:
+        """
+        Sets the value of :py:attr:`labelCol`.
+        """
+        return self.set_params(labelCol=value)
+
+
+class UMAP(UMAPClass, _CumlEstimatorSupervised, _UMAPCumlParams):
+
+    """
+    >>> from spark_rapids_ml.umap import UMAP
+    >>> X, _ = (1000, 10, centers=42, cluster_std=0.1,
+                dtype=np.float32, random_state=10)
+    >>> df = spark.createDataFrame(X, ["features"])
+    >>> local_model = UMAP().setFeaturesCol("features")
+    >>> distributed_model = umap.fit(df)
+    >>> embeddings = distributed_model.transform(df)
+    >>> embeddings.show()
+    """
+
+    def __init__(self, sample_fraction: float = 1.0, **kwargs: Any) -> None:
+        super().__init__()
+        self.set_params(**kwargs)
+        self.sample_fraction = sample_fraction
+
+    def _create_pyspark_model(self, result: Row) -> "UMAPModel":
+        return UMAPModel.from_row(result)
+
+    def _fit(self, dataset: DataFrame) -> "UMAPModel":
+        if self.sample_fraction < 1.0:
+            data_subset = dataset.sample(
+                withReplacement=False,
+                fraction=self.sample_fraction,
+                seed=self.cuml_params["random_state"],
+            )
+        else:
+            data_subset = dataset
+
+        input_num_workers = self.num_workers
+        # Force to single partition, single worker
+        self._num_workers = 1
+        if data_subset.rdd.getNumPartitions() != 1:
+            data_subset = data_subset.coalesce(1)
```

Comment:
Have you tried this case? I'm not quite sure if coalesce can work with the following barrier RDD.